core/store/src/shard_uid_mapping.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{
+    flat::POISONED_LOCK_ERR,
+    io::{Error, Result},
+};
+use near_primitives::shard_layout::ShardUId;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+/// Stores a mapping from ShardUId to ShardUId.
+///
+/// Protected with mutex for concurrent access.
+/// That is for resharding V3 purposes, where we use the mapping strategy for State column.
+#[derive(Clone)]
+pub struct ShardUIdMapping(Arc<Mutex<ShardUIdMappingInner>>);
```

Comment:
Wondering whether it makes more sense to use `Rwlock` instead of `Mutex` here? We would mostly be reading from the map and not writing.